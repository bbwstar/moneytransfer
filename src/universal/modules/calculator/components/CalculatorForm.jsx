import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import SelectField from 'components/Form/SelectField';
import TextField from 'components/Form/TextField';
import convertToSlug from 'utils/url/convertToSlug';

class CalculatorForm extends Component {
  state = {
    from: '',
    to: '',
    amount: '1000',
    type: 'sent',
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { from, to, type, amount } = this.state;
    const fromSlug = convertToSlug(from);
    const toSlug = convertToSlug(to);
    const typeSlug = convertToSlug(type);
    const amountSlug = convertToSlug(amount);

    this.props.changePage(`/${typeSlug}-money/${fromSlug}/${toSlug}/${amountSlug}`);
  };

  handleChange = (event) => {
    const { value, name } = event.target;

    this.setState({
      [name]: value,
    });
  };

  render() {
    const countriesOptions = [{ name: 'Czech Republic' }, { name: 'United Kingdom' }];
    const typeOptions = [{ name: 'Send' }, { name: 'Receive' }];

    return (
      <div className="columns">
        <div className="column is-8 is-offset-2">
          <form onSubmit={this.handleSubmit}>
            <div className="columns">
              <div className="column">
                <SelectField
                  name="From"
                  nameFormattedMessage={<FormattedMessage id="from" defaultMessage="From" />}
                  handleChange={this.handleChange}
                  value={this.state.from}
                  options={countriesOptions}
                />
              </div>
              <div className="column">
                <SelectField
                  name="To"
                  nameFormattedMessage={<FormattedMessage id="to" defaultMessage="To" />}
                  handleChange={this.handleChange}
                  value={this.state.to}
                  options={countriesOptions}
                />
              </div>
            </div>
            <div className="columns">
              <div className="column">
                <SelectField
                  name="Type"
                  nameFormattedMessage={<FormattedMessage id="type" defaultMessage="Type" />}
                  handleChange={this.handleChange}
                  value={this.state.type}
                  options={typeOptions}
                />
              </div>
              <div className="column">
                <TextField
                  name="Amount"
                  nameFormattedMessage={<FormattedMessage id="amount" defaultMessage="Amount" />}
                  handleChange={this.handleChange}
                  value={this.state.amount}
                />
              </div>
            </div>
            <div className="columns">
              <div className="column is-2 is-offset-5 has-text-centered">
                <div className="tile is-parent">
                  <p className="control">
                    <button className="button is-primary is-large is-fullwidth">
                      <FormattedMessage id="submit" defaultMessage="Submit" />
                    </button>
                  </p>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

CalculatorForm.propTypes = {
  changePage: PropTypes.func.isRequired,
};

export default CalculatorForm;
