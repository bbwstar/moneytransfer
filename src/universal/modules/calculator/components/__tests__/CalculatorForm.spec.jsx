import React from 'react';
import toJson from 'enzyme-to-json';
import { mountWithIntl } from 'utils/intl-enzyme-test-helper';

import CalculatorForm from '../CalculatorForm';

let wrapper;
const changePage = jest.fn();

describe('CalculatorForm (Component)', () => {
  beforeEach(() => {
    wrapper = mountWithIntl(<CalculatorForm changePage={changePage} />);
  });

  it('renders without exploding', () => {
    expect(wrapper).toHaveLength(1);
  });

  it('match the snaphost', () => {
    const tree = toJson(wrapper);

    expect(tree).toMatchSnapshot();
  });

  it('changes value of from select', () => {
    expect(wrapper.find('#from').props().value).toBe('');
    wrapper.find('#from').simulate('change', { target: { value: 'From new value', name: 'from' } });

    expect(wrapper.find('#from').props().value).toBe('From new value');
  });

  it('changes value of to select', () => {
    expect(wrapper.find('#to').props().value).toBe('');
    wrapper.find('#to').simulate('change', { target: { value: 'To new value', name: 'to' } });

    expect(wrapper.find('#to').props().value).toBe('To new value');
  });

  it('changes value of type select', () => {
    expect(wrapper.find('#type').props().value).toBe('sent');
    wrapper.find('#type').simulate('change', { target: { value: 'Type new value', name: 'type' } });

    expect(wrapper.find('#type').props().value).toBe('Type new value');
  });

  it('changes value of amount text field', () => {
    expect(wrapper.find('#amount').props().value).toBe('1000');
    wrapper.find('#amount').simulate('change', { target: { value: '50', name: 'amount' } });

    expect(wrapper.find('#amount').props().value).toBe('50');
  });

  describe('submit the form', () => {
    it('redirects to result page after submit', () => {
      wrapper.setState({ from: 'Italy', to: 'Czech republic', type: 'Send', amount: '1000' });

      wrapper.find('button').simulate('submit');

      expect(changePage).toBeCalledWith('/send-money/italy/czech-republic/1000');
    });

    it('prevents default action onSubmit', () => {
      const e = { preventDefault: jest.fn() };
      wrapper.find('button').simulate('submit', e);

      expect(e.preventDefault).toBeCalled();
    });
  });
});
