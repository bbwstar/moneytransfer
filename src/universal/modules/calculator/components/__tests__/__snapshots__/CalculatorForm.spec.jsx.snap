// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorForm (Component) match the snaphost 1`] = `
<CalculatorForm
  changePage={[Function]}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
    }
  }
>
  <div
    className="columns"
  >
    <div
      className="column is-8 is-offset-2"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="columns"
        >
          <div
            className="column"
          >
            <SelectField
              handleChange={[Function]}
              name="From"
              nameFormattedMessage={
                <FormattedMessage
                  defaultMessage="From"
                  id="from"
                  values={Object {}}
                />
              }
              options={
                Array [
                  Object {
                    "name": "Czech Republic",
                  },
                  Object {
                    "name": "United Kingdom",
                  },
                ]
              }
              value=""
            >
              <div
                className="field"
              >
                <label
                  className="label"
                  htmlFor="From"
                >
                  <FormattedMessage
                    defaultMessage="From"
                    id="from"
                    values={Object {}}
                  >
                    <span>
                      From
                    </span>
                  </FormattedMessage>
                </label>
                <p
                  className="control"
                >
                  <span
                    className="select is-medium is-fullwidth"
                  >
                    <select
                      id="from"
                      name="from"
                      onChange={[Function]}
                      value=""
                    >
                      <option />
                      <option
                        value="Czech Republic"
                      >
                        Czech Republic
                      </option>
                      <option
                        value="United Kingdom"
                      >
                        United Kingdom
                      </option>
                    </select>
                  </span>
                </p>
              </div>
            </SelectField>
          </div>
          <div
            className="column"
          >
            <SelectField
              handleChange={[Function]}
              name="To"
              nameFormattedMessage={
                <FormattedMessage
                  defaultMessage="To"
                  id="to"
                  values={Object {}}
                />
              }
              options={
                Array [
                  Object {
                    "name": "Czech Republic",
                  },
                  Object {
                    "name": "United Kingdom",
                  },
                ]
              }
              value=""
            >
              <div
                className="field"
              >
                <label
                  className="label"
                  htmlFor="To"
                >
                  <FormattedMessage
                    defaultMessage="To"
                    id="to"
                    values={Object {}}
                  >
                    <span>
                      To
                    </span>
                  </FormattedMessage>
                </label>
                <p
                  className="control"
                >
                  <span
                    className="select is-medium is-fullwidth"
                  >
                    <select
                      id="to"
                      name="to"
                      onChange={[Function]}
                      value=""
                    >
                      <option />
                      <option
                        value="Czech Republic"
                      >
                        Czech Republic
                      </option>
                      <option
                        value="United Kingdom"
                      >
                        United Kingdom
                      </option>
                    </select>
                  </span>
                </p>
              </div>
            </SelectField>
          </div>
        </div>
        <div
          className="columns"
        >
          <div
            className="column"
          >
            <SelectField
              handleChange={[Function]}
              name="Type"
              nameFormattedMessage={
                <FormattedMessage
                  defaultMessage="Type"
                  id="type"
                  values={Object {}}
                />
              }
              options={
                Array [
                  Object {
                    "name": "Send",
                  },
                  Object {
                    "name": "Receive",
                  },
                ]
              }
              value="sent"
            >
              <div
                className="field"
              >
                <label
                  className="label"
                  htmlFor="Type"
                >
                  <FormattedMessage
                    defaultMessage="Type"
                    id="type"
                    values={Object {}}
                  >
                    <span>
                      Type
                    </span>
                  </FormattedMessage>
                </label>
                <p
                  className="control"
                >
                  <span
                    className="select is-medium is-fullwidth"
                  >
                    <select
                      id="type"
                      name="type"
                      onChange={[Function]}
                      value="sent"
                    >
                      <option />
                      <option
                        value="Send"
                      >
                        Send
                      </option>
                      <option
                        value="Receive"
                      >
                        Receive
                      </option>
                    </select>
                  </span>
                </p>
              </div>
            </SelectField>
          </div>
          <div
            className="column"
          >
            <TextField
              handleChange={[Function]}
              name="Amount"
              nameFormattedMessage={
                <FormattedMessage
                  defaultMessage="Amount"
                  id="amount"
                  values={Object {}}
                />
              }
              value="1000"
            >
              <div
                className="field"
              >
                <label
                  className="label"
                  htmlFor="amount"
                >
                  <FormattedMessage
                    defaultMessage="Amount"
                    id="amount"
                    values={Object {}}
                  >
                    <span>
                      Amount
                    </span>
                  </FormattedMessage>
                </label>
                <p
                  className="control"
                >
                  <input
                    className="input is-medium is-fullwidth"
                    id="amount"
                    name="amount"
                    onChange={[Function]}
                    type="text"
                    value="1000"
                  />
                </p>
              </div>
            </TextField>
          </div>
        </div>
        <div
          className="columns"
        >
          <div
            className="column is-2 is-offset-5 has-text-centered"
          >
            <div
              className="tile is-parent"
            >
              <p
                className="control"
              >
                <button
                  className="button is-primary is-large is-fullwidth"
                >
                  <FormattedMessage
                    defaultMessage="Submit"
                    id="submit"
                    values={Object {}}
                  >
                    <span>
                      Submit
                    </span>
                  </FormattedMessage>
                </button>
              </p>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</CalculatorForm>
`;
